solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+		diag.range.start = _file.buffer().toPosition(pos);
+		diag.range.end = {diag.range.start.line, diag.range.start.column + 5};
+		diag.severity = DiagnosticSeverity::Hint;
+		diag.source = "solc";
+		params.diagnostics.emplace_back(diag);
+	}
+#endif
+
+	_result.emplace_back(params);
+}
+
+frontend::ASTNode const* LanguageServer::findASTNode(lsp::Position const& _position, std::string const& _fileName)
+{
+	if (!m_compilerStack)
+		return nullptr;
+
+	frontend::ASTNode const& sourceUnit = m_compilerStack->ast(_fileName);
+	auto const sourcePos = sourceUnit.location().source->translateLineColumnToPosition(_position.line + 1, _position.column + 1);
+
+	ASTNodeLocator m{sourcePos};
+	sourceUnit.accept(m);
+	auto const closestMatch = m.closestMatch();
+	if (closestMatch != nullptr)
+		fprintf(stderr, "findASTNode for %s @ pos=%d:%d (%d), symbol: '%s' (%s)\n",
+				sourceUnit.location().source->name().c_str(),
+				sourcePos,
+				_position.line,
+				_position.column,
+				closestMatch->location().text().c_str(), typeid(*closestMatch).name());
+	else
+		fprintf(stderr, "findASTNode for pos=%d:%d (%d), not found.\n",
+				sourcePos,
+				_position.line,
+				_position.column);
+
+	return closestMatch;
+}
+
+optional<LanguageServer::Location> LanguageServer::gotoDefinition(DocumentPosition _location)
+{
+	if (auto const file = m_vfs.find(_location.uri); file != nullptr)
+	{
+		compile(*file);
+		solAssert(m_compilerStack.get() != nullptr, "");
+
+		auto const sourceName = file->uri().substr(7); // strip "file://"
+
+		if (auto const sourceNode = findASTNode(_location.position, sourceName); sourceNode)
+		{
+			if (auto const importDirective = dynamic_cast<ImportDirective const*>(sourceNode))
+			{
+				// When cursor is on an import directive, then we want to jump to the actual file that
+				// is being imported.
+				if (auto const fpm = m_fileReader->fullPathMapping().find(importDirective->path()); fpm != m_fileReader->fullPathMapping().end())
+				{
+					Location output{};
+					output.uri = "file://" + fpm->second;
+					return {output};
+				}
+				else
+					return nullopt; // definition not found
+			}
+			else if (auto const declaration = dynamic_cast<Declaration const*>(sourceNode))
+			{
+				// For any kind of declaration, jump to the referencing symbol of that declaration.
+				if (auto const loc = declarationPosition(declaration); loc.has_value())
+				{
+					Location output{};
+					output.range = loc.value();
+					output.uri = "file://" + declaration->location().source->name();
+					return output;
+				}
+				else
+					return nullopt; // definition not found
+			}
+			else if (auto const n = dynamic_cast<frontend::MemberAccess const*>(sourceNode); n)
+			{
+				// For scope members, jump to the naming symbol of the referencing declaration of this member.
+				auto const declaration = n->annotation().referencedDeclaration;
+
+				if (auto const loc = declarationPosition(declaration); loc.has_value())
```

Comment:
Duplicated code.

Reply:
duplicated with what? Maybe I'm blind right now ;-(

Reply:
several lines up or down - everything from the point where you have a declaration.